add(state): Backup non-finalized blocks in state cache directory

arya2 opened the pull request:
## Motivation

We want Zebra to recover its non-finalized state when it restarts.

Closes #9708.

## Solution

- Creates a `{state_cache_dir}/non_finalized_state/{network.lowercase_name()}` directory
- Adds a config field to allow for disabling non-finalized state backups
- Attempts to commit blocks from the non-finalized backup cache into the non-finalized state when constructing the `StateService`
- Listens to the non-finalized state receiver and:
  - Writes any blocks in the non-finalized state that are missing in the backup cache to the backup cache, and
  - Deletes any files in the backup cache that are not present in the non-finalized state.


Related changes:
- Checks the non-finalized state for a block candidate's parent block before checking RocksDB

### Tests

This PR adds a unit test and has been manually tested.

### Follow-up Work

Add a config field to enable/disable the non-finalized state backup cache.

### PR Checklist

<!-- Check as many boxes as possible. -->

- [x] The PR name is suitable for the release notes.
- [x] The solution is tested.
- [x] The documentation is up to date.

arya2 committed:
add(state): Backup non-finalized state in state cache directory

- Defines a non-finalized state backup directory path relative to the state cache directory path
- Adds a `NonFinalizedState::restore_backup()` method that reads from the directory and attempts to commit any blocks in it

arya2 committed:
- Moves logic for restoring backups to its own module
- Refactors `restore_backup()` method into `with_backup()` method that should also spawn the backup task

arya2 committed:
- Moves logic for reading and processing backup files out of `restore_backup()` into their own fns
- Adds and calls a stub `spawn_backup_task()` fn for writing the latest non-finalized state to the backup dir, and
- Updates `with_backup()` method docs.

arya2 committed:
- Makes fns in `backup` mod private to the `non_finalized_state` mod
- Adds an info log
- Adds a `backup_task()` fn

arya2 committed:
Makes `StateService::new()` an async fn

arya2 committed:
Updates `duration_short()` to print microseconds when formatting durations less than 5 ms

arya2 committed:
- Moves blocking ops in `StateService::new()` to blocking tasks, and
- Updates warning msg from `NonFinalizedBlocksListener::spawn()` when non-finalized state receiver is closed.

arya2 committed:
- Checks that block hashes in file name match computed block hashes when restoring blocks from non-finalized state backup
- Removes unnecessary function

arya2 committed:
Implements `run_backup_task()`, writing blocks in the non-finalized state to the backup cache.

arya2 committed:
- Fixes a bug where the non-finalized state would not be restored from the backup cache
- Fixes a bug where the state service uses the finalized tip as the initial chain tip in the latest chain tip channels even when there are non-finalized blocks restored from the backup cache

arya2 commented on zebra-chain/src/fmt/time.rs:
```diff
@@ -42,3 +47,17 @@ pub fn humantime_milliseconds(duration: impl Into<Duration>) -> String {
     format!("{duration}")
 }
+
+/// Returns a human-friendly formatted string for the whole number of milliseconds in `duration`.
```
```suggestion
/// Returns a human-friendly formatted string for the whole number of microseconds in `duration`.
```

(These changes were made while checking for places where `spawn_blocking()` may have been needed since `StateService::new()` has been made into an async method)

arya2 committed:
Rate limits updates to non-finalized state backup cache

arya2 committed:
Adds a config field to allow for disabling non-finalized state backup cache.

arya2 committed:
Fixes clippy lints, minor cleanup

arya2 committed:
- Reads/writes deferred pool balance change to non-finalized block backup files
- Adds a unit test

arya2 committed:
Updates `validate_and_commit_non_finalized()` to check for parent block in non-finalized state before checking rocksdb

arya2 commented on zebra-state/src/service/non_finalized_state/backup.rs:
```diff
@@ -0,0 +1,326 @@
+    }
+}
+
+impl NonFinalizedBlockBackup {
```
We could also compute the deferred pool balance change instead of reading/writing it. 

Let me know if that's preferrable.

```rust
// Calculates the [`DeferredPoolBalanceChange`] of a post-NU6 transaction by subtracting the total
// value balance of all of its outputs from its coinbase input value.
//
// > From the activation block of this ZIP onward, coinbase transactions MUST claim all of the available
// > miner subsidy plus fees in their block.
// <https://zips.z.cash/zip-0236>, ZIP 236: Blocks should balance exactly
// <https://zips.z.cash/zip-0253>, ZIP 253: Deployment of the NU6 Network Upgrade
//
// > • A coinbase transaction MUST NOT have any JoinSplit descriptions.
// > • A coinbase transaction MUST NOT have any Spend descriptions.
// > • [Pre-Heartwood] A coinbase transaction MUST NOT have any Output descriptions.
// > • [ NU5onward] In a version 5 coinbase transaction, the enableSpendsOrchard flag MUST be 0
// <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>, section 7.1.2, Transaction Consensus Rules (p. 122)
```

arya2 committed:
Fixes clippy lints

arya2 committed:
Fixes bad test config

arya2 commented:
This PR will need to update tests using `future_blocks()` or `raw_future_blocks()`, currently the `rpc_submit_block()` and `send_raw_transaction()` tests.

Update: Fixed in https://github.com/ZcashFoundation/zebra/pull/9809/commits/6d8dcf9b2f27ebff93848ab8ee3c4b04378f4299

arya2 committed:
Updates `raw_future_blocks()` to avoid caching a backup of its non-finalized state on disk

oxarbitrage commented on zebra-state/src/service/write.rs:
```diff
@@ -57,7 +57,9 @@ pub(crate) fn validate_and_commit_non_finalized(
     let parent_hash = prepared.block.header.previous_block_hash;
 
-    if finalized_state.finalized_tip_hash() == parent_hash {
+    if !non_finalized_state.any_chain_contains(&parent_hash)
```
Did you spotted an issue here or is just extra checking ?

arya2 replied:
It's just an extra check to avoid disk I/O

oxarbitrage commented:
I tried to add this test: https://github.com/ZcashFoundation/zebra/commit/ec1faa31534eba68fa34a9736e46f6efbc8651e7 with this branch and i think there is an issue after you recover the non finalized state. It seems the chain can't make progress after that. Can you take a look ?

Also, can you paste some logs of your manual testing ? Thanks.

arya2 committed:
Drops finalized block write sender when instantiating state service if the finalized tip height is above the max checkpoint height, or skips restoring non-finalized state from backup cache otherwise.

arya2 committed:
Adds any blocks that were restored from the non-finalized backup cache to `SentHashes` when instantiating StateService

arya2 committed:
updates test config

arya2 committed:
Fixes a bug where the non-finalized state backup directory would not be created if the finalized tip is below the max checkpoint height

arya2 committed:
Fixes a bug where Zebra would log the wrong initial state cache height when blocks are restored from the non-finalized state backup cache

arya2 committed:
Merge remote-tracking branch 'origin/main' into non-final-blocks-backup

arya2 committed:
Merge remote-tracking branch 'origin/main' into non-final-blocks-backup

arya2 committed:
Merge remote-tracking branch 'origin/main' into non-final-blocks-backup

oxarbitrage commented:
I have an additional test for this. Here is the code, let me know if you will like me to push it as a commit here after https://github.com/ZcashFoundation/zebra/pull/9888 gets merged or if you want me to do it in a separated PR after https://github.com/ZcashFoundation/zebra/pull/9809 gets merged.

zebra-rpc/qa/rpc-tests/feature_backup_non_finalized_state.py:

```python
#!/usr/bin/env python3
# Copyright (c) 2025 The Zcash developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or https://www.opensource.org/licenses/mit-license.php .

from decimal import Decimal
from test_framework.util import assert_equal, start_nodes
from test_framework.test_framework import BitcoinTestFramework
import time

# Test that Zebra can backup and restore non finalized state
class BackupNonFinalized(BitcoinTestFramework):

    def __init__(self):
        super().__init__()
        self.num_nodes = 1
        self.cache_behavior = 'clean'

    def setup_network(self):
        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, [None])

    def run_test(self):
        self.nodes[0].generate(10)
        # Wait for `MIN_DURATION_BETWEEN_BACKUP_UPDATES`, 5 secs in zebra
        time.sleep(5)

        blocks = self.nodes[0].getblockchaininfo()['blocks']
        assert_equal(blocks, 10)

        # Stop the node
        self.nodes[0].stop()
        time.sleep(1)

        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, [None])

        # Good, the node has recovered the non finalized state
        blocks = self.nodes[0].getblockchaininfo()['blocks']
        assert_equal(blocks, 10)

        # Generate more blocks and make sure the blockchain is still working
        self.nodes[0].generate(1)
        time.sleep(5)

        blocks = self.nodes[0].getblockchaininfo()['blocks']
        assert_equal(blocks, 11)

if __name__ == '__main__':
    BackupNonFinalized().main()
```

Also, we need to get rid of the hack introduced at https://github.com/ZcashFoundation/zebra/blob/main/zebra-rpc/qa/rpc-tests/wallet.py#L51 to bypass the limitation. We can do it in a separated PR.

arya2 commented:
> I have an additional test for this. Here is the code, let me know if you will like me to push it as a commit here after #9888 gets merged or if you want me to do it in a separated PR after #9809 gets merged.
> 
> Also, we need to get rid of the hack introduced at https://github.com/ZcashFoundation/zebra/blob/main/zebra-rpc/qa/rpc-tests/wallet.py#L51 to bypass the limitation. We can do it in a separated PR.

Could you open a suggestion PR against this one with both the new test and the removal of the workaround in the existing test? We could also include it in a follow-up PR to minimize the changes in this one.

It seems the chain stall issue may not have been fixed, as all of [the integration tests on this PR are still timing out](https://github.com/ZcashFoundation/zebra/actions/runs/17671838575/job/50227052558?pr=9809#step:13:337), even though the new acceptance test in https://github.com/ZcashFoundation/zebra/pull/9888 is passing.

Update: Found and fixed the third and seemingly final issue with the chain stall in https://github.com/ZcashFoundation/zebra/pull/9809/commits/b33136736d3721a61b45bfe22a11850c8564fbba. ~The acceptance test added in the follow up PR #9888 doesn't yet cover it.~

Update: Added test coverage for the last issue in https://github.com/ZcashFoundation/zebra/pull/9888/commits/88af388bdd9a3aed1baf3b891c7b11f2042da71e.

oxarbitrage commented:
Yes, i opened https://github.com/ZcashFoundation/zebra/pull/9896

But based on https://github.com/ZcashFoundation/zebra/pull/9888

arya2 committed:
Updates state service initialization to read the best tip in a blocking task

arya2 committed:
Adds a note in the field docs for `should_backup_non_finalized_state` about rate limits

arya2 committed:
Fixes a bug where the chain would stall when the finalized tip is past the max checkpoint hash and Zebra is attempting to restore blocks from the non-finalized backup cache, but commits no valid blocks to the initial non-finalized state and initializes the finalized block write sender as `None` such that `sent_hashes.can_fork_chain_at` always remains false.
